src/node/txdownloadman_impl.cpp

```diff
@@ -0,0 +1,523 @@
+            // because we should not download or submit this transaction by itself again, but may
+            // submit it as part of a package later.
+            RecentRejectsReconsiderableFilter().insert(ptx->GetWitnessHash().ToUint256());
+
+            if (first_time_failure) {
+                // When a transaction fails for TX_RECONSIDERABLE, look for a matching child in the
+                // orphanage, as it is possible that they succeed as a package.
+                LogDebug(BCLog::TXPACKAGES, "tx %s (wtxid=%s) failed but reconsiderable, looking for child in orphanage\n",
+                         ptx->GetHash().ToString(), ptx->GetWitnessHash().ToString());
+                package_to_validate = Find1P1CPackage(ptx, nodeid);
+            }
+        } else {
+            RecentRejectsFilter().insert(ptx->GetWitnessHash().ToUint256());
+        }
+        m_txrequest.ForgetTxHash(ptx->GetWitnessHash());
+        // If the transaction failed for TX_INPUTS_NOT_STANDARD,
+        // then we know that the witness was irrelevant to the policy
+        // failure, since this check depends only on the txid
+        // (the scriptPubKey being spent is covered by the txid).
+        // Add the txid to the reject filter to prevent repeated
+        // processing of this transaction in the event that child
+        // transactions are later received (resulting in
+        // parent-fetching by txid via the orphan-handling logic).
+        // We only add the txid if it differs from the wtxid, to avoid wasting entries in the
+        // rolling bloom filter.
+        if (state.GetResult() == TxValidationResult::TX_INPUTS_NOT_STANDARD && ptx->HasWitness()) {
+            RecentRejectsFilter().insert(ptx->GetHash().ToUint256());
+            m_txrequest.ForgetTxHash(ptx->GetHash());
+        }
+    }
+
+    // If the tx failed in ProcessOrphanTx, it should be removed from the orphanage unless the
+    // tx was still missing inputs. If the tx was not in the orphanage, EraseTx does nothing and returns 0.
+    if (state.GetResult() != TxValidationResult::TX_MISSING_INPUTS && m_orphanage.EraseTx(ptx->GetWitnessHash()) > 0) {
+        LogDebug(BCLog::TXPACKAGES, "   removed orphan tx %s (wtxid=%s)\n", ptx->GetHash().ToString(), ptx->GetWitnessHash().ToString());
+    }
+
+    return RejectedTxTodo{
+        .m_should_add_extra_compact_tx = add_extra_compact_tx,
+        .m_unique_parents = std::move(unique_parents),
+        .m_package_to_validate = std::move(package_to_validate)
+    };
+}
+
+void TxDownloadManagerImpl::MempoolRejectedPackage(const Package& package)
+{
+    RecentRejectsReconsiderableFilter().insert(GetPackageHash(package));
+}
+
+std::pair<bool, std::optional<PackageToValidate>> TxDownloadManagerImpl::ReceivedTx(NodeId nodeid, const CTransactionRef& ptx)
+{
+    const uint256& txid = ptx->GetHash();
+    const uint256& wtxid = ptx->GetWitnessHash();
```

Comment:
Why is it that we don't initialize these with the`Txid` and `Wtxid` types? Have been thinking about making `RelayTransaction` type safe, so just wondering.

Reply:
txrequest and the bloom filters take uint256s, so this seemed easiest. I don't feel strongly. `RelayTransaction` isn't called here...?